### PR TITLE
[JIT] Make aot autograd decompositions usable in JIT, add script for serializing the decompositions

### PR DESCRIPTION
**Pull request by eellison**
Stack from [ghstack](https://github.com/ezyang/ghstack):
* #74348
* #74076
* #74186
* #74012
* __->__ #73938
* #73876
* #73875

This is a first step in porting and making usable all of the decompositions defined in [functorch](https://github.com/pytorch/functorch/blob/main/functorch/_src/decompositions.py#L349) in core and in JIT as well as C++. 

The decompositions are defined in python, scripted and inlined, and then serialized as C++ code which TorchScript can parse. The workflow is edit python decomposition file then run [tools/codegen/decompositions/gen_jit_decompositions.py](https://github.com/pytorch/pytorch/pull/73938/files#diff-6adef2116be233c3524e3b583e373ab0ffc9169beb6c1f6d96b5d0385e75afa1). 

Decompositions are mapped to their corresponding aten schemas via the schema in their python def. This allows multiple decompositions for an overloaded op like `aten.var` (shown here in the example). 

This is just a first PR, i'm sure there will be many follows ups such as:
- making these runnable in C++ with simple executor
- porting over more decompositions from AOT Autograd
- Using opinfos / more robust testing
- Categorizing decompositions 
- Hooking in decompositions at various points of JIT execution

Differential Revision: [D34938126](https://our.internmc.facebook.com/intern/diff/D34938126)

### CODE REVIEW

File: torch/csrc/jit/runtime/decomposition_registry.h [hunk 1]
```diff
@@ -0,0 +1,17 @@
...
+#include <torch/csrc/Export.h>
+#include <torch/csrc/jit/ir/ir.h>
+
+namespace torch {
+namespace jit {
+
+TORCH_API c10::optional<std::shared_ptr<Graph>> DecompositionGraphForSchema(
+    const FunctionSchema& schema);
+
+TORCH_API void RunDecompositions(std::shared_ptr<Graph> g);
```
Comment (Chillee) [inline torch/csrc/jit/runtime/decomposition_registry.h:14]: Would be good to have these allow selective use of decompositions.

**Review by Chillee — APPROVED** [review]
LGTM - modulo some of the comments we discussed offline.

File: tools/codegen/decompositions/gen_jit_decompositions.py [hunk 2]
```diff
@@ -0,0 +1,83 @@
...
+    with open(
+        os.path.join(path, DECOMPOSITION_UTIL_FILE_NAME), "wb"
+    ) as out_file:
+        final_output = "".join(file_components)
+        out_file.write(final_output.encode("utf-8"))
+
+def main() -> None:
+    pytorch_dir = Path(__file__).resolve().parents[3]
+    upgrader_path = pytorch_dir / "torch" / "csrc" / "jit" / "runtime"
+    write_decomposition_util_file(str(upgrader_path))
```
[the remark this answers is not in the document]
Comment (eellison) [reply tools/codegen/decompositions/gen_jit_decompositions.py:79]: There are like 5 shared lines.. i'll look into refactoring

[the remark this answers is not in the document]
Comment (Chillee) [reply tools/codegen/decompositions/gen_jit_decompositions.py:79]: If there's not that much started, dw about it.

**Comment by eellison** [comment]
@eellison has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D34938126).

**Comment by eellison** [comment]
@eellison has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D34938126).

**Comment by eellison** [comment]
@eellison has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D34938126).

**Comment by eellison** [comment]
@eellison has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D34938126).

**Comment by eellison** [comment]
@eellison has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D34938126).

**Comment by eellison** [comment]
@eellison has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D34938126).

**Comment by eellison** [comment]
@eellison has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D34938126).

**Comment by eellison** [comment]
@eellison has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D34938126).

**Pull request closed without merging**